driver/src/orderbook/retrying_retrieval.rs

```diff
@@ -0,0 +1,260 @@
+}
+
+fn read_orderbook(
+    contract: &dyn StableXContractAsync,
+    batch_id: u32,
+    page_size: u16,
+) -> Result<(AccountState, Vec<Order>)> {
+    futures::executor::block_on(async {
+        // NOTE: Use a bounded channel with a 0 sized buffer so that the sender
+        //   (i.e. the future that is polling the incomming events) can only
+        //   send 1 retry at a time.
+        let (retry_tx, retry_rx) = mpsc::channel::<()>(0);
+        let (done_tx, done_rx) = oneshot::channel::<()>();
+
+        let (solution_event, latest_solution, (account_state, orders)) = try_join!(
+            poll_incoming_events(contract, batch_id, retry_tx, done_rx),
+            contract.latest_solution(batch_id),
+            read_orderbook_async(contract, batch_id, page_size, retry_rx, done_tx),
+        )?;
+
+        if let Some((solution, trades)) = solution_event.or(latest_solution) {
+            todo!("undo {:?} and from {:?} from orderbook", solution, trades);
+        }
+
+        Ok((account_state, orders))
+    })
+}
+
+async fn read_orderbook_async(
+    contract: &dyn StableXContractAsync,
+    batch_id: u32,
+    page_size: u16,
+    mut retry: mpsc::Receiver<()>,
+    _done: oneshot::Sender<()>,
+) -> Result<(AccountState, Vec<Order>)> {
+    let mut reader = PaginatedAuctionDataReader::new(batch_id.into());
+    loop {
+        let mut page_future = contract
+            .encoded_orders_paginated(
+                reader.pagination().previous_page_user,
+                reader
+                    .pagination()
+                    .previous_page_user_offset
+                    .try_into()
+                    .expect("user cannot have more than u16::MAX orders"),
+                page_size,
+            )
+            .fuse();
+
+        let page = select! {
+            page = page_future => page?,
+            _ = retry.next() => {
+                reader = PaginatedAuctionDataReader::new(batch_id.into());
+                continue;
+            },
+            complete => break Err(anyhow!("error processing incoming events")),
+        };
+
+        let number_of_orders: u16 = reader
+            .apply_page(&page)
+            .try_into()
+            .expect("number of orders per page should never overflow a u16");
+
+        if number_of_orders < page_size {
+            return Ok(reader.get_auction_data());
+        }
+    }
+
+    // NOTE: `done` gets dropped automatically when this method returns, making
+    //   its `Receiver` resolve to `Err(Cancelled)` which will cause the event
+    //   loop to break.
+}
+
+async fn poll_incoming_events(
+    contract: &dyn StableXContractAsync,
+    batch_id: u32,
+    mut retry: mpsc::Sender<()>,
+    done: oneshot::Receiver<()>,
+) -> Result<Option<SolutionData>> {
+    use batch_exchange::Event::*;
+    use Batch::*;
+    use EventData::*;
+
+    let mut done = done.fuse();
+    let mut events = contract.all_events().fuse();
```

Comment:
Is this giving us all events from the beginning of time or just since we call this one? I assume the first, as we want to  get the latest solution. But this means we would retry the fetching initially a lot of times as we iterate over all past OrderCancellations.

Maybe we should query the current block ID, base all our fetch queries on that block and only restart (with a new blocknumber) if there is an order cancellation newer than our starting block but still in the past batch.

Reply:
> Maybe we can limit from when we fetch events to something like 20 blocks in the past

I think this is a really good idea. We currently bail if we get an event in the next batch, but we can also listen on block number.

> and make the invalidation logic a bit less verbose

Do you mean the logic that determines if a retry is needed?

> Is this giving us all events from the beginning of time or just since we call this one?

It is from the call onwards, we have a separate future getting the last applied solution that we join on. We take the latest solution (from incoming events) or the latest solution from the aforementioned future. I did this to avoid the issue you just mentioned.

> Maybe we should query the current block ID, base all our fetch queries on that block

I was under the impression we ran into issues with this because of state pruning.

Reply:
> Do you mean the logic that determines if a retry is needed?

Makes sense if we are only getting events from the point of subscription onwards. I originally thought we are querying events from genesis.